### PR TITLE
[#IOPAE-749] Azure access token refresh on expiration

### PR DESCRIPTION
**Pull request by giamma1295**
**Add Azure Access Token refresh on expiration capabilities**
#### List of Changes
- Edit `getApimRestClient` function, now will refresh the token on 401 or 403 apim response.
- Add `azure-access-token.ts` module which provides the Azure Access Token

#### How Has This Been Tested?
Unit test

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by rodoviti — COMMENTED** [review]
There could be a moment so `now > expires`, but when the request will actually be executed, the token may be expired (i.g.: when `now` _is almost equals to_ `expires`).
In this case, the request will fails.

I suggest to mange the token refresh by handling the 403 (I think) APIM error response. In this way, we will refresh token during a retry after a 403 response, do you agree?

File: apps/backoffice/src/lib/be/__tests__/apim-service.test.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+      expect(get).toHaveBeenCalledOnce();
+      expect(E.isRight(result)).toBeTruthy();
+      if (E.isRight(result)) {
+        expect(result.right).toStrictEqual({ name: mocks.aServiceId });
+      }
+    });
+
+    it("should retry when apim response is 401 or 403 and return the list", async () => {
+      const anExpiredAzureAccessToken = "anExpiredAzureAccessToken";
+      const aNotExpiredAzureAccessToken = "aNotExpiredAzureAccessToken";
```
[the remark this answers is not in the document]
Comment (giamma1295) [reply apps/backoffice/src/lib/be/__tests__/apim-service.test.ts:94]: This is a test module, no api call are made here, everything is mocked..

File: apps/backoffice/src/lib/be/__tests__/apim-service.test.ts [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+      expect(get).toHaveBeenCalledOnce();
+      expect(E.isRight(result)).toBeTruthy();
+      if (E.isRight(result)) {
+        expect(result.right).toStrictEqual({ name: mocks.aServiceId });
+      }
+    });
+
+    it("should retry when apim response is 401 or 403 and return the list", async () => {
+      const anExpiredAzureAccessToken = "anExpiredAzureAccessToken";
```
[the remark this answers is not in the document]
Comment (giamma1295) [reply apps/backoffice/src/lib/be/__tests__/apim-service.test.ts:93]: This is a test module, no api call are made here, everything is mocked..

File: apps/backoffice/src/lib/be/apim-service.ts [hunk 3]
```diff
@@ -49,42 +52,38 @@ export const getApimService: () => ApimUtils.ApimService = cache(() => {
   );
 });
 
-const getAxiosInstance = async () => {
+const getAxiosInstance = cache((azureAccessToken: string) => {
```
Comment (rodoviti) [inline apps/backoffice/src/lib/be/apim-service.ts:55]: **chore:** now cache is no more needed here

File: apps/backoffice/src/lib/be/azure-access-token.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  return tokenResponse;
+};
+
+export const getAzureAccessToken = async (): Promise<string> => {
+  // Lazy init accessToken
+  if (!accessToken) {
+    accessToken = await refreshAzureAccessToken();
+    console.log(
```
Comment (rodoviti) [inline apps/backoffice/src/lib/be/azure-access-token.ts:67]: **suggestion:** log this event as debug level

File: apps/backoffice/src/lib/be/azure-access-token.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        accessToken.expiresOnTimestamp
+      )}`
+    );
+  } else {
+    // Lazy Referesh accessToken
+    const now = new Date();
+    const expires = new Date(accessToken.expiresOnTimestamp);
+    if (now > expires) {
+      accessToken = await refreshAzureAccessToken();
+      console.log(
```
Comment (rodoviti) [inline apps/backoffice/src/lib/be/azure-access-token.ts:78]: **suggestion:** log this event as debug level

File: apps/backoffice/src/lib/be/azure-access-token.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+      apimConfig.AZURE_CLIENT_SECRET_CREDENTIAL_SECRET,
+      {
+        authorityHost: AzureAuthorityHosts.AzurePublicCloud
+      }
+    );
+    tokenResponse = await credential.getToken(
+      apimConfig.AZURE_CREDENTIALS_SCOPE_URL
+    );
+  } catch (e) {
+    console.log("Error retrieving on Azure Access Token", e);
```
Comment (rodoviti) [inline apps/backoffice/src/lib/be/azure-access-token.ts:52]: **suggestion:** log this event as error level

Comment (giamma1295) [reply apps/backoffice/src/lib/be/azure-access-token.ts]: done.
BTW this module has been removed

File: apps/backoffice/src/lib/be/azure-access-token.ts (hunk 5 above)
Comment (giamma1295) [reply apps/backoffice/src/lib/be/azure-access-token.ts]: this module has been removed and this functionality was moved back on `apim-service.ts`

File: apps/backoffice/src/lib/be/azure-access-token.ts (hunk 4 above)
Comment (giamma1295) [reply apps/backoffice/src/lib/be/azure-access-token.ts]: module removed

File: apps/backoffice/src/lib/be/apim-service.ts (hunk 3 above)
Comment (giamma1295) [reply apps/backoffice/src/lib/be/apim-service.ts]: done

Comment (giamma1295) [reply apps/backoffice/src/lib/be/apim-service.ts:55]: needed in change after review

File: apps/backoffice/src/lib/be/azure-access-token.ts (hunk 4 above)
Comment (giamma1295) [reply apps/backoffice/src/lib/be/azure-access-token.ts]: done

File: apps/backoffice/src/lib/be/azure-access-token.ts (hunk 5 above)
Comment (giamma1295) [reply apps/backoffice/src/lib/be/azure-access-token.ts]: done

File: apps/backoffice/src/lib/be/azure-access-token.ts (hunk 6 above)
Comment (giamma1295) [reply apps/backoffice/src/lib/be/azure-access-token.ts]: done

**Review by giuseppedipinto — APPROVED** [review]

**Pull request merged**